File: projects/yukkuri-remotion/voicevox_gen.py
import requests
import json
import os
import time

def generate_voice(text, speaker_id, output_path):
    print(f"Generating audio for: {text[:20]}...")
    
    # 1. Create Audio Query
    response = requests.post(
        f"http://localhost:50021/audio_query",
        params={"text": text, "speaker": speaker_id}
    )
    if response.status_code != 200:
        print(f"Error creating audio query: {response.text}")
        return False
    
    query = response.json()
    
    # 2. Synthesis
    response = requests.post(
        f"http://localhost:50021/synthesis",
        params={"speaker": speaker_id},
        data=json.dumps(query)
    )
    if response.status_code != 200:
        print(f"Error synthesis: {response.text}")
        return False
    
    with open(output_path, "wb") as f:
        f.write(response.content)
    
    return True

def main():
    # Load dialogue
    data_path = "projects/yukkuri-remotion/video/src/data.json"
    with open(data_path, "r") as f:
        dialogue = json.load(f)
    
    audio_dir = "projects/yukkuri-remotion/video/public/audio"
    os.makedirs(audio_dir, exist_ok=True)
    
    # Speaker IDs (Standard: Zundamon=3, Metan=2)
    speakers = {
        "zundamon": 3,
        "metan": 2
    }
    
    for i, item in enumerate(dialogue):
        speaker_id = speakers.get(item["speaker"], 3)
        output_file = f"voice_{i}.wav"
        output_path = os.path.join(audio_dir, output_file)
        
        success = generate_voice(item["text"], speaker_id, output_path)
        if success:
            item["audio"] = output_file
        else:
            print(f"Failed to generate audio for line {i}")
            
    # Update data.json with actual audio filenames
    with open(data_path, "w") as f:
        json.dump(dialogue, f, indent=2, ensure_ascii=False)
    
    print("Audio generation complete!")

if __name__ == "__main__":
    # Wait for service to be ready
    for _ in range(10):
        try:
            requests.get("http://localhost:50021/version")
            print("VOICEVOX service is ready!")
            break
        except:
            print("Waiting for VOICEVOX...")
            time.sleep(5)
    
    main()

File: projects/yukkuri-remotion/voicevox_gen.js
const fs = require('fs');
const path = require('path');

const API_URL = 'http://localhost:50021';

async function generateVoice(text, speakerId, outputPath) {
    console.log(`Generating audio for: ${text.substring(0, 20)}...`);
    
    // 1. Create Audio Query
    const queryUrl = `${API_URL}/audio_query?text=${encodeURIComponent(text)}&speaker=${speakerId}`;
    const queryResponse = await fetch(queryUrl, { method: 'POST' });
    if (!queryResponse.ok) {
        console.error(`Error creating audio query: ${await queryResponse.text()}`);
        return false;
    }
    const query = await queryResponse.json();
    
    // 2. Synthesis
    const synthesisUrl = `${API_URL}/synthesis?speaker=${speakerId}`;
    const synthesisResponse = await fetch(synthesisUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(query)
    });
    if (!synthesisResponse.ok) {
        console.error(`Error synthesis: ${await synthesisResponse.text()}`);
        return false;
    }
    
    const buffer = Buffer.from(await synthesisResponse.arrayBuffer());
    fs.writeFileSync(outputPath, buffer);
    return true;
}

async function main() {
    const dataPath = path.join(__dirname, 'video/src/data.json');
    const dialogue = JSON.parse(fs.readFileSync(dataPath, 'utf8'));
    
    const audioDir = path.join(__dirname, 'video/public/audio');
    if (!fs.existsSync(audioDir)) {
        fs.mkdirSync(audioDir, { recursive: true });
    }
    
    const speakers = {
        zundamon: 3,
        metan: 2
    };
    
    for (let i = 0; i < dialogue.length; i++) {
        const item = dialogue[i];
        const speakerId = speakers[item.speaker] || 3;
        const outputFile = `voice_${i}.wav`;
        const outputPath = path.join(audioDir, outputFile);
        
        const success = await generateVoice(item.text, speakerId, outputPath);
        if (success) {
            item.audio = outputFile;
        } else {
            console.error(`Failed to generate audio for line ${i}`);
        }
    }
    
    fs.writeFileSync(dataPath, JSON.stringify(dialogue, null, 2));
    console.log('Audio generation complete!');
}

async function waitForService() {
    for (let i = 0; i < 20; i++) {
        try {
            const response = await fetch(`${API_URL}/version`);
            if (response.ok) {
                console.log('VOICEVOX service is ready!');
                return true;
            }
        } catch (e) {
            console.log('Waiting for VOICEVOX...');
            await new Promise(resolve => setTimeout(resolve, 5000));
        }
    }
    return false;
}

(async () => {
    if (await waitForService()) {
        await main();
    } else {
        console.error('VOICEVOX service timed out.');
    }
})();
